### PR TITLE
feat(wc): duplicate orders admin notice

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adds an admin notice when duplicate orders are detected:

image

### How to test the changes in this Pull Request:

1. Donate two times with the same amount, within 10 minutes
2. Do that again, but change the dates of these two to more than a month ago
3. Trigger the CRON job: `wp cron event run newspack_wc_check_order_duplicates`
4. Visit the admin panel, observe the admin notice about potentially duplicated orders – from the last month (the first two created)
5. Dismiss one of the cases, observe the page reloads and the list item is no longer displayed
6. Run `wp newspack detect-order-duplicates --cutoff_time='1 year' --save` and refresh the admin page
7. Observe three items are present on the list: two new ones from more than a month ago

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208711277186607

### CODE REVIEW

**Review by leogermani — COMMENTED** [review]
Even though the SQL queries are pretty efficient, they are huge, hard to read and hard to maintain.

And also hard to test, because I need two different sites to test it.

Doing this with simple code, looping though recent orders would be less efficient, but much more readable, maintainable and testable... And considering that this check runs daily, we don't need to look at all the orders in the history, just the most recent ones.

**Comment by adekbadek** [comment]
Yeah, that makes sense. I've updated with the following:

1. Instead of SQL query, iterate order objects retrieved with `wc_get_orders` (in batches of 100, so the site does not fall over)
2. Added a `wp newspack detect-order-duplicates` CLI command to enable filling in historical data (older than a month). This will have to be ran on live sites after the PR is deployed.

Update the PR description accordingly.

File: includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+			$dismissed_duplicates[] = $_POST['dismiss_order_ids']; // phpcs:ignore WordPress.Security.NonceVerification.Missing, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+			update_option( self::DISMISSED_DUPLICATES_OPTION_NAME, $dismissed_duplicates );
+			// Refresh the page to reflect changes.
+			wp_safe_redirect( isset( $_SERVER['REQUEST_URI'] ) ? $_SERVER['REQUEST_URI'] : admin_url() );
+			exit;
+		}
+	}
+
+	/**
+	 * CLI handler to upsert the order duplicates with a specified timeframe.
```
Comment (leogermani) [inline includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php]: is it detect or upsert?

File: includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	 *
+	 * [--cutoff_time=<time-string>]
+	 * : The cutoff time in the past (e.g. "2 months").
+	 *
+	 * [--save]
+	 * : Whether to save the results for display in the admin panel.
+	 *
+	 * ## EXAMPLES
+	 *
+	 *     wp newspack detect-order-duplicates --cutoff_time='2 months' --save
```
Comment (leogermani) [inline includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php:230]: is it detect or upsert?

File: includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php (hunk 1 above)
Comment (leogermani) [reply includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php]: Let's add a clearer description of what the command does
- search for duplicates and optionally store this info to be displayed in the admin panel

**Review by leogermani — COMMENTED** [review]
Works as described! Thanks for the changes! I left a few minor comments.

I just noticed one thing. I first ran the CLI command without the `--save` param. It detected the duplicated but didn't save. However, it ended up displaying the admin notice, but with no orders listed. Just the summary. When you clicked, it was an empty list.

Another question: Since we're running this daily, and cutting off based on completion date, why default to check the last 30 days? Looks like we will be doing a lot of unnecessary checks every day. Isn't it enough to check just the last one or two days?

File: includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php (hunk 2 above)
Comment (adekbadek) [reply includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php:230]: Without `--save`, it's only detection. With it, it's upsertion.

File: includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php (hunk 1 above)
Comment (adekbadek) [reply includes/reader-revenue/woocommerce/class-woocommerce-duplicate-orders.php]: Updated in 520f49c1367f00aebf272734c855943de2c7080a

**Comment by adekbadek** [comment]
> However, it ended up displaying the admin notice, but with no orders listed.

Yeah, I failed to account for an empty state, thanks for catching that! Fixed in 520f49c1367f00aebf272734c855943de2c7080a

> Isn't it enough to check just the last one or two days?

That's right, changed to a day in 520f49c1367f00aebf272734c855943de2c7080a

**Review by leogermani — APPROVED** [review]

**Pull request merged**